ofrak_core/ofrak/gui/script_builder.py

```diff
@@ -0,0 +1,280 @@
+        session.hashed_actions[session.actions_counter] = ScriptAction(action_type, action)
+        session.actions_counter += 1
+
+    async def _add_variable_to_session(self, resource: Resource, var_name: str):
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource.get_id())
+        session.variable_mapping[resource.get_id()] = var_name
+
+    async def _get_variable_from_session(self, resource: Resource):
+        root_resource = await self._get_root_resource(resource)
+        return self.script_sessions[root_resource.get_id()].variable_mapping[resource.get_id()]
+
+    async def _var_exists(self, resource: Resource):
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource)
+        return resource.get_id() in session.variable_mapping
+
+    def delete_action(self, resource_id: bytes, action: str) -> None:
+        """
+        :param action:
+        """
+        # TODO: do we really need to delete an action from the script?
+        for key, script_action in self.script_sessions[resource_id].hashed_actions.items():
+            if script_action.action == action:
+                del self.script_sessions[resource_id].hashed_actions[key]
+
+    def _get_session(self, resource_id: bytes) -> ScriptSession:
+        session = self.script_sessions.get(resource_id, None)
+        if session is None:
+            session = ScriptSession()
+            self.script_sessions[resource_id] = session
+
+        return session
+
+    async def get_script(self, resource: Resource) -> str:
+        """
+        :return script:
+        """
+        root_resource = await self._get_root_resource(resource)
+        return self._get_script(root_resource.get_id())
+
+    def _get_script(self, resource_id: bytes, target_type: Optional[ActionType] = None) -> str:
+        script = []
+        script.append(self.script_sessions[resource_id].boilerplate_header)
+        for script_action in self.script_sessions[resource_id].hashed_actions.values():
+            if target_type is None or target_type == script_action.action_type:
+                script.append(f"{script_action.action}")
+        script.append(self.script_sessions[resource_id].boilerplate_footer)
+        script = "\n".join(script)
+        script = self._dedent(script)
+        res = format_str("\n".join(script), mode=FileMode())
+        script = res.split("\n")
+        return script
+
+    def _dedent(self, s):
+        split = list(s.splitlines())
+        prefix = os.path.commonprefix([line for line in split if line != ""])
+        indent_matches = re.findall(r"^\s+", prefix)
+        if not indent_matches:
+            return s
+        indent_end_index = len(indent_matches[0])
+        return [line[indent_end_index:] for line in split if line.startswith(prefix) and line != ""]
+
+    def get_all_of_type(self, resource_id: bytes, target_type: ActionType) -> str:
```

Comment:
Unused

Reply:
The intention is for this to be used when we need to support selectively creating scripts based on action types (e.g., "get all steps that involved unpacking down to this resource so that I can repeat this unpacking process again").